### PR TITLE
Remove the use of common buffer

### PR DESCRIPTION
**Pull request by AniruddhaKanhere**
This PR removes the use of common buffer used for creating and transmitting packets. The reception will be handled in the user provided buffer.

### CODE REVIEW

**Review by paulbartell — APPROVED** [review]

File: source/core_mqtt.c [hunk 1]
```diff
@@ -1532,56 +1758,318 @@ static MQTTStatus_t validateSubscribeUnsubscribeParams( const MQTTContext_t * pC
...
-        LogError( ( "Transport send failed for PUBLISH header." ) );
         status = MQTTSendFailed;
     }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static void addWillAndConnectInfo( const MQTTConnectInfo_t * pConnectInfo,
```
Comment (paulbartell) [inline source/core_mqtt.c:1948]: addWillAndConnectInfo is only called from sendConnectWithoutCopy. Can we incorporate addWillAndConnectInfo into that function instead?

File: source/core_mqtt.c [hunk 2]
```diff
@@ -1532,56 +1758,318 @@ static MQTTStatus_t validateSubscribeUnsubscribeParams( const MQTTContext_t * pC
...
+                                   TransportOutVector_t * iterator,
+                                   uint8_t serializedTopicLength[ 2 ],
+                                   uint8_t serializedPayloadLength[ 2 ],
+                                   uint8_t serializedUsernameLength[ 2 ],
+                                   uint8_t serializedPasswordLength[ 2 ] )
+{
+    if( pWillInfo != NULL )
+    {
+        /* Serialize the topic. */
+        iterator = addEncodedStringToVector( serializedTopicLength,
```
Comment (paulbartell) [inline source/core_mqtt.c:1960]: I'm not sure that this improves clarity really. Wouldn't it be simpler just to fill the array?

File: source/core_mqtt.c [hunk 3]
```diff
@@ -2026,31 +2493,11 @@ MQTTStatus_t MQTT_Subscribe( MQTTContext_t * pContext,
...
-            LogError( ( "Transport send failed for SUBSCRIBE packet." ) );
-            status = MQTTSendFailed;
-        }
-        else
-        {
-            LogDebug( ( "Sent %ld bytes of SUBSCRIBE packet.",
-                        ( long int ) bytesSent ) );
-        }
+        /* Send MQTT SUBSCRIBE packet. */
+        status = sendSubscribeWithoutCopy( pContext,
```
Comment (paulbartell) [inline source/core_mqtt.c:2497]: Does sendSubscribeWithoutCopy need to be a separate function?

File: source/core_mqtt.c (hunk 1 above)
Comment (AniruddhaKanhere) [reply source/core_mqtt.c:2006]: Doing that increases the complexity score.

**Review by paulbartell — DISMISSED** [review]

File: source/include/core_mqtt_serializer.h [hunk 4]
```diff
@@ -76,6 +76,116 @@
...
+ *
+ * @param[in] x The unsigned int to check.
+ * @param[in] position Which bit to check.
+ */
+#define UINT8_CHECK_BIT( x, position )    ( ( ( x ) & ( 0x01U << ( position ) ) ) == ( 0x01U << ( position ) ) )
+
+/**
+ * @brief Get the high byte of a 16-bit unsigned integer.
+ */
+#define UINT16_HIGH_BYTE( x )             ( ( uint8_t ) ( ( x ) >> 8 ) )
```
Comment (archigup) [inline source/include/core_mqtt_serializer.h:157]: These should not be in a public header

File: source/interface/transport_interface.h [hunk 5]
```diff
@@ -243,15 +243,54 @@ typedef int32_t ( * TransportSend_t )( NetworkContext_t * pNetworkContext,
...
+
+    /**
+     * @brief Length of data in buffer.
+     */
+    size_t iov_len;
+} TransportOutVector_t;
+
+/**
+ * @transportcallback
+ * @brief Transport interface function for "vectored" / scatter-gather based writes.
```
Comment (archigup) [inline source/interface/transport_interface.h:264]: It is unclear what someone implementing this function has to make it do.

File: source/include/core_mqtt.h [hunk 6]
```diff
@@ -58,13 +58,19 @@
 #define MQTT_LIBRARY_VERSION    "v1.2.0"
 /** @endcond */
 
+/**
+ * @ingroup mqtt_constants
+ * @brief Maximum number of vectors in subscribe and unsubscribe packet.
+ */
+#define MQTT_SUB_UNSUB_MAX_VECTORS    ( 4U )
```
Comment (archigup) [inline source/include/core_mqtt.h:65]: This seems unused?

Comment (AniruddhaKanhere) [reply source/include/core_mqtt.h:65]: This is used in the function `sendSubscribeWithoutCopy`.

Comment (archigup) [reply source/include/core_mqtt.h:65]: Ahh my diff was missing a file, makes more sense now

File: source/interface/transport_interface.h [hunk 7]
```diff
@@ -243,15 +243,54 @@ typedef int32_t ( * TransportSend_t )( NetworkContext_t * pNetworkContext,
...
+ *
+ * @param[in] pNetworkContext Implementation-defined network context.
+ * @param[in] pIoVec An array of TransportIoVector_t structs.
+ * @param[in] ioVecCount Number of TransportIoVector_t in pIoVec.
+ *
+ * @return The number of bytes written or a negative value to indicate error.
+ *
+ * @note If no data is written to the buffer due to the buffer being full this MUST
+ * return zero as the return value.
+ * A zero return value SHOULD represent that the send operation can be retried
```
[the remark this answers is not in the document]
Comment (AniruddhaKanhere) [reply source/interface/transport_interface.h]: Updated

File: source/include/core_mqtt_serializer.h (hunk 4 above)
Comment (AniruddhaKanhere) [reply source/include/core_mqtt_serializer.h]: Fixed

**Review by archigup — APPROVED** [review]

**Review by paulbartell — APPROVED** [review]

**Pull request merged**